### PR TITLE
Escape sequences can now be set to a custom value

### PR DESCRIPTION
**Pull request by swaathi**
While most ANSI color coding uses the `\033` escape sequence, some consoles might set a custom value. This PR introduces `Ansispan.escape_sequence` which can be set to any custom value.

Example:
```
irb(main):001:0" content = "\\u001b[31mF\\u001b[0m
irb(main):002:0" 
irb(main):003:0" Failures:
irb(main):004:0" 
irb(main):005:0"   1) addition adds two numbers
irb(main):006:0"      \\u001b[31mFailure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2, 3)).to eq(7)   end end\\u001b[0m
irb(main):007:0"      \\u001b[31m\\u001b[0m
irb(main):008:0"      \\u001b[31m  expected: 7\\u001b[0m
irb(main):009:0"      \\u001b[31m       got: 5\\u001b[0m
irb(main):010:0"      \\u001b[31m\\u001b[0m
irb(main):011:0"      \\u001b[31m  (compared using ==)\\u001b[0m
irb(main):012:0"      \\u001b[36m# /tmp/test.rb:11:in `block (2 levels) in <main>'\\u001b[0m
irb(main):013:0" 
irb(main):014:0" Finished in 0.00711 seconds (files took 0.04915 seconds to load)
irb(main):015:0" \\u001b[31m1 example, 1 failure\\u001b[0m
irb(main):016:0" 
irb(main):017:0" Failed examples:
irb(main):018:0" 
irb(main):019:0" \\u001b[31mrspec /tmp/test.rb:11\\u001b[0m \\u001b[36m# addition adds two numbers\\u001b[0m"
=> "\\u001b[31mF\\u001b[0m\n\nFailures:\n\n  1) addition adds two numbers\n     \\u001b[31mFailure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2,...
irb(main):020:0> Ansispan.escape_sequence = "\\u001b"
=> "\\u001b"
irb(main):021:0> Ansispan.convert(content)
=> "<span style=\"color: red\">F</span>\n\nFailures:\n\n  1) addition adds two numbers\n     <span style=\"color: red\">Failure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2, 3)).to eq(7)   end end</span>\n     <span style=\"color: red\"></span>\n     <span style=\"color: red\">  expected: 7</span>\n     <span style=\"color: red\">       got: 5</span>\n     <span style=\"color: red\"></span>\n     <span style=\"color: red\">  (compared using ==)</span>\n     <span style=\"color: cyan\"># /tmp/test.rb:11:in `block (2 levels) in <main>'</span>\n\nFinished in 0.00711 seconds (files took 0.04915 seconds to load)\n<span style=\"color: red\">1 example, 1 failure</span>\n\nFailed examples:\n\n<span style=\"color: red\">rspec /tmp/test.rb:11</span> <span style=\"color: cyan\"># addition adds two numbers</span>"
irb(main):022:0>
```

### CODE REVIEW

**Review by jelaniwoods — COMMENTED** [review]
I tried [the example you posted](https://github.com/firstdraft/learn/issues/620#issuecomment-2090423067) in the console but it didn't appear to work correctly

image

Does this work for you?

I noticed also that `\u001b` and `\033` are the same?

image

Or maybe that's only because of my console settings 🤔 I don't know a lot about ansi codes 😅

**Comment by swaathi** [comment]
Ah I should have posted an example here. Here you go,

```
irb(main):001:0" content = "\\u001b[31mF\\u001b[0m
irb(main):002:0" 
irb(main):003:0" Failures:
irb(main):004:0" 
irb(main):005:0"   1) addition adds two numbers
irb(main):006:0"      \\u001b[31mFailure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2, 3)).to eq(7)   end end\\u001b[0m
irb(main):007:0"      \\u001b[31m\\u001b[0m
irb(main):008:0"      \\u001b[31m  expected: 7\\u001b[0m
irb(main):009:0"      \\u001b[31m       got: 5\\u001b[0m
irb(main):010:0"      \\u001b[31m\\u001b[0m
irb(main):011:0"      \\u001b[31m  (compared using ==)\\u001b[0m
irb(main):012:0"      \\u001b[36m# /tmp/test.rb:11:in `block (2 levels) in <main>'\\u001b[0m
irb(main):013:0" 
irb(main):014:0" Finished in 0.00711 seconds (files took 0.04915 seconds to load)
irb(main):015:0" \\u001b[31m1 example, 1 failure\\u001b[0m
irb(main):016:0" 
irb(main):017:0" Failed examples:
irb(main):018:0" 
irb(main):019:0" \\u001b[31mrspec /tmp/test.rb:11\\u001b[0m \\u001b[36m# addition adds two numbers\\u001b[0m"
=> "\\u001b[31mF\\u001b[0m\n\nFailures:\n\n  1) addition adds two numbers\n     \\u001b[31mFailure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2,...
irb(main):020:0> Ansispan.escape_sequence = "\\\\u001b"
=> "\\\\u001b"
irb(main):021:0> Ansispan.convert(content)
=> "<span style=\"color: red\">F</span>\n\nFailures:\n\n  1) addition adds two numbers\n     <span style=\"color: red\">Failure/Error: describe 'addition' do   it 'adds two numbers' do     expect(add(2, 3)).to eq(7)   end end</span>\n     <span style=\"color: red\"></span>\n     <span style=\"color: red\">  expected: 7</span>\n     <span style=\"color: red\">       got: 5</span>\n     <span style=\"color: red\"></span>\n     <span style=\"color: red\">  (compared using ==)</span>\n     <span style=\"color: cyan\"># /tmp/test.rb:11:in `block (2 levels) in <main>'</span>\n\nFinished in 0.00711 seconds (files took 0.04915 seconds to load)\n<span style=\"color: red\">1 example, 1 failure</span>\n\nFailed examples:\n\n<span style=\"color: red\">rspec /tmp/test.rb:11</span> <span style=\"color: cyan\"># addition adds two numbers</span>"
irb(main):022:0>
```

Since there are two `\\` you'll need to escape each one so the escape sequence becomes `\\\\u001b`.

I think `\033` is the octal representation and ﻿`\u001b` is the Unicode representation of the same `ESC` ASCII character. But since we do a string search here in the UTF8 format (I think?) these two "characters" are different.

**Comment by swaathi** [comment]
Pushed an update that escapes the escape character 😅

Your earlier example of `Ansispan.escape_sequence = "\\u001b"` should work now.

File: lib/ansispan.rb [hunk 1]
```diff
@@ -1,4 +1,14 @@
 class Ansispan
+  @@escape_sequence = '\033'
+
+  def self.escape_sequence
+    @@escape_sequence
+  end
+
+  def self.escape_sequence=(custom_escape_sequence)
+    @@escape_sequence = Regexp.escape(custom_escape_sequence)
+  end
```
Comment (jelaniwoods) [inline lib/ansispan.rb]: I feel like the most common usage of this gem will just be using it for one conversion and not using it multiple times (unless maybe you're in a loop). With that in mind, do you think it makes more sense to add `escape_sequence` as an optional param to `Ansispan.convert`?

```rb
def self.convert(str, escape_character: '\033')
  ...
end

# ...

Ansispan.convert(output, escape_character: '\001b')
```

**Review by jelaniwoods — APPROVED** [review]
> Since there are two `\\` you'll need to escape each one so the escape sequence becomes `\\\\u001b`.

Ah, of course! Though I wonder why this escape character is escaped with a `\` in the first place? I've never seen it happen for other escape characters like `\033` or `\e` 🤔 

> I think \033 is the octal representation and ﻿\u001b is the Unicode representation of the same ESC ASCII character. But since we do a string search here in the UTF8 format (I think?) these two "characters" are different.

This sounds vaguely familiar... did you happen to know any good references for ANSI/octal/unicode? I don't think I saved any from when I first worked on this.

LGTM, this works as expected. I left one more comment, you're free to not address in you disagree. I don't want to block your progress on firstdraft/learn#620 by this.

**Comment by swaathi** [comment]
> did you happen to know any good references for ANSI/octal/unicode? I don't think I saved any from when I first worked on this

Nothing solid, just a bunch of Stackoverflow threads and this [great gist](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797).

File: lib/ansispan.rb (hunk 1 above)
Comment (swaathi) [reply lib/ansispan.rb]: I like this! Will make a quick edit.

**Pull request merged**